File: prueba-tecnica-backend/src/main/java/com/lautaro_cenizo/service/TecnicoService.java
package com.lautaro_cenizo.service;

import com.lautaro_cenizo.crud.dto.JornadaLaboralDTO;
import com.lautaro_cenizo.crud.dto.TecnicoDTO;
import com.lautaro_cenizo.entity.historial.HistorialServicio;
import com.lautaro_cenizo.entity.tecnico.DiasSemana;
import com.lautaro_cenizo.entity.tecnico.Especialidad;
import com.lautaro_cenizo.entity.tecnico.Tecnico;

import java.time.LocalDateTime;
import java.util.List;
import java.util.Map;

public interface TecnicoService {
    Tecnico crear(TecnicoDTO tecnicoDTO);
    Tecnico actualizar(Integer id, TecnicoDTO tecnicoDTO);
    Tecnico obtenerPorId(Integer id);
    Tecnico obtenerPorCardId(Long cardId);
    List<Tecnico> listarTodos();
    List<Tecnico> listarActivos();
    List<Tecnico> listarPorEspecialidad(Especialidad especialidad);
    void eliminar(Integer id);
    void registrarServicio(Integer tecnicoId, HistorialServicio servicio);
    boolean verificarDisponibilidad(Integer tecnicoId, LocalDateTime fecha);
    void actualizarJornadaLaboral(Integer tecnicoId, Map<DiasSemana, JornadaLaboralDTO> jornadaLaboral);
}

File: prueba-tecnica-frontend/src/app/service/tecnico.service.ts
import { Injectable } from '@angular/core';
import { TecnicoDTO } from '../models/dto/tecnicoDto';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { Tecnico } from '../models/class/tecnico';
import { Especialidad } from '../models/enums/especialidad';
import { HistorialServicio } from '../models/class/historialServicio';
import { DiasSemana } from '../models/enums/diaSemana';
import { JornadaLaboralDTO } from '../models/dto/jornadaLaboralDto';

@Injectable({
  providedIn: 'root'
})
export class TecnicoService {

  private baseUrl = 'http://localhost:8080/api/tecnicos';

  constructor(private http: HttpClient) { }

  crear(tecnicoDTO: TecnicoDTO): Observable<Tecnico> {
    return this.http.post<Tecnico>(this.baseUrl, tecnicoDTO);
  }

  actualizar(id: number, tecnicoDTO: TecnicoDTO): Observable<Tecnico> {
    return this.http.put<Tecnico>(`${this.baseUrl}/${id}`, tecnicoDTO);
  }

  obtenerPorId(id: number): Observable<Tecnico> {
    return this.http.get<Tecnico>(`${this.baseUrl}/${id}`);
  }

  obtenerPorCardId(cardId: number): Observable<Tecnico> {
    return this.http.get<Tecnico>(`${this.baseUrl}/card/${cardId}`);
  }

  listarTodos(): Observable<Tecnico[]> {
    return this.http.get<Tecnico[]>(this.baseUrl);
  }

  listarActivos(): Observable<Tecnico[]> {
    return this.http.get<Tecnico[]>(`${this.baseUrl}/activos`);
  }

  listarPorEspecialidad(especialidad: Especialidad): Observable<Tecnico[]> {
    return this.http.get<Tecnico[]>(`${this.baseUrl}/especialidad/${especialidad}`);
  }

  eliminar(id: number): Observable<void> {
    return this.http.delete<void>(`${this.baseUrl}/${id}`);
  }

  registrarServicio(tecnicoId: number, servicio: HistorialServicio): Observable<void> {
    return this.http.post<void>(`${this.baseUrl}/${tecnicoId}/registrar-servicio`, servicio);
  }

  verificarDisponibilidad(tecnicoId: number, fecha: Date): Observable<boolean> {
    return this.http.get<boolean>(
      `${this.baseUrl}/${tecnicoId}/disponibilidad`,
      { params: { fecha: fecha.toISOString() } }
    );
  }

  actualizarJornadaLaboral(
    tecnicoId: number,
    jornadaLaboral: Record<DiasSemana, JornadaLaboralDTO>
  ): Observable<void> {
    return this.http.put<void>(
      `${this.baseUrl}/${tecnicoId}/jornada-laboral`,
      jornadaLaboral
    );
  }
}
